eksupgrade/models/eks.py

```diff
@@ -0,0 +1,357 @@
+"""Define the models to be used across the EKS upgrade tool."""
+from __future__ import annotations
+
+import base64
+import datetime
+import logging
+from abc import ABC
+from dataclasses import dataclass, field
+from functools import cached_property
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union
+
+import boto3
+from kubernetes import client as k8s_client
+from kubernetes import config as k8s_config
+
+from .base import AwsRegionResource
+
+try:
+    from functools import cache
+except ImportError:  # pragma: no cover
+    from functools import lru_cache as cache
+
+if TYPE_CHECKING:  # pragma: no cover
+    from mypy_boto3_eks.client import EKSClient
+    from mypy_boto3_eks.type_defs import (
+        AddonInfoTypeDef,
+        AddonTypeDef,
+        ClusterTypeDef,
+        DescribeAddonResponseTypeDef,
+        DescribeClusterResponseTypeDef,
+        UpdateAddonResponseTypeDef,
+    )
+else:
+    AddonInfoTypeDef = object
+    AddonTypeDef = object
+    ClusterTypeDef = object
+    DescribeAddonResponseTypeDef = object
+    DescribeClusterResponseTypeDef = object
+    EKSClient = object
+    UpdateAddonResponseTypeDef = object
+
+logger = logging.getLogger(__name__)
+
+TOKEN_PREFIX: str = "k8s-aws-v1"
+TOKEN_HEADER_KEY: str = "x-k8s-aws-id"
+
+
+def requires_cluster(function):
+    """Decorate methods to require a cluster attribute."""
+
+    def wrapper(self, *args, **kwargs):
+        if not self.cluster.name:
+            logger.error(
+                "Unable to use method: %s without the cluster attribute! Pass a cluster to this child object!",
+                function.__name__,
+            )
+            return None
+        return function(self, *args, **kwargs)
+
+    return wrapper
+
+
+@dataclass
+class EksResource(AwsRegionResource, ABC):
+    """Define the abstract EKS base resource class."""
+
+    name: str = ""
+    status: str = ""
+    version: str = ""
+
+    @cached_property
+    def eks_client(self) -> EKSClient:
+        """Get a boto EKS client."""
+        return self._get_boto_client(service="eks", region_name=self.region)
+
+    @cached_property
+    def core_api_client(self) -> Any:
+        """Get a Kubernetes Core client."""
+        return k8s_client.CoreV1Api()
+
+    @cached_property
+    def apps_api_client(self) -> Any:
+        """Get a Kubernetes Apps client."""
+        return k8s_client.AppsV1Api()
+
+
+@dataclass
+class ClusterAddon(EksResource):
+    """Define the Kubernetes Cluster Addon model."""
+
+    health: Dict[str, List[Any]] = field(default_factory=lambda: ({"issues": []}))
+    created_at: datetime.datetime = datetime.datetime.now()
```

Comment:
any particular reason why we are setting the timestamp as opposed to using an empty string and populating with the values returned by the API?